cdap-ui/app/hydrator/services/hydrator-upgrade-service.js

```diff
@@ -0,0 +1,254 @@
+        data.error = 'SCOPE_MISMATCH';
+        data.suggestion = pluginsMap[stageKey].highestVersion;
+      }
+
+      transformedStages.push(data);
+    });
+
+    return transformedStages;
+  }
+
+  upgradePipelineArtifactVersion(pipelineConfig) {
+    if (!pipelineConfig || !pipelineConfig.artifact) { return; }
+
+    let cdapVersion = this.$rootScope.cdapVersion;
+
+    let configClone = _.cloneDeep(pipelineConfig);
+
+    configClone.artifact.version = cdapVersion;
+
+    return configClone;
+  }
+
+  validateAndUpgradeConfig(pipelineConfig) {
+    this.$uibModal.open({
+      templateUrl: '/assets/features/hydrator/templates/create/pipeline-upgrade-modal.html',
+      size: 'lg',
+      backdrop: 'static',
+      keyboard: false,
+      windowTopClass: 'hydrator-modal center upgrade-modal',
+      controllerAs: 'PipelineUpgradeController',
+      controller: function ($scope, rPipelineConfig, HydratorUpgradeService, $rootScope, HydratorPlusPlusConfigStore, $state, DAGPlusPlusFactory, GLOBALS, HydratorPlusPlusLeftPanelStore) {
+        let eventEmitter = window.CaskCommon.ee(window.CaskCommon.ee);
+        let globalEvents = window.CaskCommon.globalEvents;
+
+        this.pipelineConfig = rPipelineConfig;
+        this.cdapVersion = $rootScope.cdapVersion;
+
+        this.pipelineArtifact = HydratorUpgradeService.checkPipelineArtifactVersion(rPipelineConfig);
+
+        this.problematicStages = [];
+        this.missingArtifactStages = [];
+        let allStages = [];
+        this.fixAllDisabled = true;
+
+        this.loading = true;
+
+        const checkStages = () => {
+          this.loading = true;
+          let transformedStages = HydratorUpgradeService.getErrorStages(rPipelineConfig);
+
+          allStages = transformedStages.map((stage) => {
+            stage.icon = DAGPlusPlusFactory.getIcon(stage.stageInfo.plugin.name.toLowerCase());
+            stage.type = GLOBALS.pluginConvert[stage.stageInfo.plugin.type];
+            return stage;
+          });
+
+          this.problematicStages = [];
+          this.missingArtifactStages = [];
+
+          transformedStages.forEach((artifact) => {
+            if (artifact.error === 'NOTFOUND') {
+              this.missingArtifactStages.push(artifact);
+            } else if (artifact.error) {
+              this.problematicStages.push(artifact);
+            }
+          });
+
+          this.fixAllDisabled = this.missingArtifactStages.length > 0;
+
+          if (this.problematicStages.length === 0 && this.missingArtifactStages.length === 0 && this.pipelineArtifact) {
+            HydratorPlusPlusConfigStore.setState(HydratorPlusPlusConfigStore.getDefaults());
+            $state.go('hydrator.create', { data: rPipelineConfig });
+          } else {
+            this.loading = false;
+          }
+        };
+
+        checkStages();
+
+        let sub = HydratorPlusPlusLeftPanelStore.subscribe(checkStages);
```

Comment:
Do we need this subscription? In `checkStages` we don't use left panel store anyways.